src/Protocol/Diaspora.php

```diff
@@ -3215,6 +3219,51 @@ private static function buildAndTransmit($owner, $contact, $type, $message, $pub
 		return $return_code;
 	}
 
+	/**
+	 * @brief sends a participation (Used to get all further updates)
+	 *
+	 * @param array $contact Target of the communication
+	 * @param array $item	 Item array
+	 *
+	 * @return int The result of the transmission
+	 */
+	private static function sendParticipation($contact, $item)
+	{
+		// Don't send notifications for private postings
+		if ($item['private']) {
+			return;
+		}
+
+		$cachekey = "diaspora:sendParticipation:".$item['guid'];
+
+		$result = Cache::get($cachekey);
+		if (!is_null($result)) {
+			return;
+		}
+
+		// Fetch some user id to have a valid handle to transmit the participation.
+		// In fact it doesn't matter which user sends this - but it is needed by the protocol.
```

Comment:
Can’t we use the `$importer` variable available in both places we call `sendParticipation` instead of the same user over and over again?

Reply:
No, we can't. Under certain circumstances the "importer" contains dummy values: https://github.com/friendica/friendica/blob/develop/src/Protocol/Diaspora.php#L487

Reply:
Especially for public posts, gotcha. Can we pass a dummy user handle then? I’m really uncomfortable with picking an arbitrary user from the user list.

Reply:
No, we need a real handle since it is checked upon receiving.

Reply:
We need a real sender, because it's needed for the signature validation (because the protocol needs it). It doesn't matter which user this is. Currently diaspora stores this user, but in the future we only store the pod after we validated the signature.

Reply:
I’m curious as to why you need to check the signature of the receiver on public post? I’m not very familiar with cryptography and thought signature was important to authenticate a sent message from its intended author.

Anyway, thanks for chiming in!

Reply:
Because all messages are wrapped inside a [MagicEnvelope](https://diaspora.github.io/diaspora_federation/federation/magicsig.html), which contains a signature which is validated before the content is processed. So we can't make an exception for participation messages here.

Even if we could receive it without signatures, it shouldn't be done. The author is now used to see which server is the sender, so we would need to add only the sender domain name instead of the full ID to the message. But without a signature, everybody can send messages to server x, and tell them, that server y is the sender. But when the message is signed with the any key from server y, we can validate, that server y is really the sender. And using a "random" users key (the first user, which is usually the admin), is an easy way to use existing stuff and does solve the problem completely.

Otherwise we would need to also create one key-pair per server, which isn't really intended by the protocol at the moment (we would also need to exchange these keys), and also doesn't have any advantages for sending participations.

Reply:
It makes all the sense in the workd, thank you so much for taking the time to explain the innervworkings of the protocol!